lib/Conversion/ExportVerilog/PrepareForEmission.cpp

```diff
@@ -697,6 +697,84 @@ static LogicalResult legalizeHWModule(Block &block,
   // True if these operations are in a procedural region.
   bool isProceduralRegion = block.getParentOp()->hasTrait<ProceduralRegion>();
 
+  // Lower HW wires into SV wires in the appropriate spots. Try to put the
+  // declaration close to the first use in the block, and the assignment right
+  // after the assigned value becomes available.
+  struct WireLowering {
+    hw::WireOp wireOp;
+    Block::iterator declarePoint;
+    Block::iterator assignPoint;
+  };
+  SmallVector<WireLowering> wireLowerings;
+
+  for (auto hwWireOp : block.getOps<hw::WireOp>()) {
+    // Find the earliest point in the block at which the input operand is
+    // available. The assign operation will have to go after that to not create
+    // a use-before-def situation.
+    Block::iterator assignPoint = block.begin();
+    if (auto *defOp = hwWireOp.getInput().getDefiningOp())
+      if (defOp && defOp->getBlock() == &block)
+        assignPoint = ++Block::iterator(defOp);
+
+    // Find the earliest use of the wire within the block. The wire declaration
+    // will have to go somewhere before this point to not create a
+    // use-before-def situation.
+    Block::iterator declarePoint = assignPoint;
+    for (auto *user : hwWireOp->getUsers()) {
+      while (user->getBlock() != &block)
+        user = user->getParentOp();
+      if (user->isBeforeInBlock(&*declarePoint))
+        declarePoint = Block::iterator(user);
+    }
+
+    wireLowerings.push_back({hwWireOp, declarePoint, assignPoint});
+  }
+
+  for (auto [hwWireOp, declarePoint, assignPoint] : wireLowerings) {
+    // Create the declaration.
+    ImplicitLocOpBuilder builder(hwWireOp.getLoc(), &block, declarePoint);
+    Value decl;
+    if (isProceduralRegion) {
+      decl = builder.create<LogicOp>(hwWireOp.getType(), hwWireOp.getNameAttr(),
+                                     hwWireOp.getInnerSymAttr());
+    } else {
+      decl =
+          builder.create<sv::WireOp>(hwWireOp.getType(), hwWireOp.getNameAttr(),
+                                     hwWireOp.getInnerSymAttr());
+    }
+
+    // Carry attributes over to the lowered operation.
+    auto defaultAttrNames = hwWireOp.getAttributeNames();
+    for (auto namedAttr : hwWireOp->getAttrs())
+      if (!llvm::is_contained(defaultAttrNames, namedAttr.getName()))
```

Comment:
This is nice, I hadn't seen this approach. I think there may also be some helper like `getDiscardableAttrs` or something along those lines that might do basically this with a little more abstraction.

Reply:
There was some [discussion on the LLVM Discourse](https://discourse.llvm.org/t/rfc-introducing-mlir-operation-properties/67846#breaking-the-model-further-opt-in-all-inherent-attributes-into-properties-8), but it doesn't seem anything has landed. It would be pretty neat though if the catch-all attribute dictionary would no longer contain the ODS-defined attributes. That would clean things up quite a bit.

Reply:
Even besides the Property idea, I think there's a way to get all discardable attributes. The method I was thinking of was `getDialectAttrs`, not sure if that is exactly what you're looking for here though.